Allow relative-to-root toctree dirs in autosummary

jakelishman opened the pull request:
Subject: Allow relative-to-root toctree dirs in autosummary

### Feature or Bugfix
- Feature

### Purpose

At the moment, Autosummary's `toctree` directory is always relative to the current file. This means that the generated files always move with the file that contains an toctree-creating `autosummary` directive. In our case, we'd like to universally put the generated files into one separate folder at the documentation root, regardless of where they were generated from. This is particularly a problem when dealing with recursive autosummary generations, and at the moment is effectively limiting us to our manual documentation all having a flat folder structure.

### Detail

This treats a leading `/` in the directory of a `toctree` option in an `autosummary` block as requesting a directory relative to the source documentation root, rather than relative to the current directory.  Previously, prefixing a path with `/` would typically cause build failures on Unix-likes because

```python
posixpath.join("/prefix/path", "/other") == "/other"
```

and Sphinx (hopefully!) didn't have permission to create directories in the filesystem root.  This means that the change should not conflict with any existing practical working documentation builds.

Autosummary currently uses `posixpath` for most operations even on Windows, so the root indicator is only permitted as `/`, not `os.sep`.


### Relates
- Fix #11637

jakelishman committed:
Allow relative-to-root toctree dirs in autosummary

This treats a leading `/` in the directory of a `toctree` option in an
`autosummary` block as requesting a directory relative to the source
documentation root, rather than relative to the current directory.
Previously, prefixing a path with `/` would typically cause build
failures on Unix-likes because

    posixpath.join("/prefix/path", "/other") == "/other"

and Sphinx (hopefully!) didn't have permission to create directories in
the filesystem root.  This means that the change should not conflict
with any existing practical working documentation builds.

Autosummary currently uses `posixpath` for most operations even on
Windows, so the root indicator is only permitted as `/`, not `os.sep`.

picnixz commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -74,13 +74,22 @@ The :mod:`sphinx.ext.autosummary` extension does this in two parts:
 
      You can also use ``caption`` option to give a caption to the toctree.
 
+     .. versionchanged:: 7.3
```
Since it's a feature, I think it's more a ``versionadded`` that is needed rather than a ``versionchanged``.

jakelishman replied:
Sure, I wasn't sure how Sphinx preferred these things.  Done in 62b27c495.

picnixz replied:
Ah no worry. Actually there isn't any strong convention. It's just that I think "versionchanged" should be reserved to "orthogonal" things changing wrt to the previous implementation without being deprecated. But anyone is free to argue !

picnixz commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -74,13 +74,22 @@ The :mod:`sphinx.ext.autosummary` extension does this in two parts:
-     as the file that contains the directive.
+     as the file that contains the directive.  The directory is interpreted as
+     relative to the directory of the source file, unless it begins with the
+     system path separator (``/`` for Unix-likes), in which case it is relative
```
Does it mean that for Windows, it should be `\` ? You shouldn't say "system path separator" otherwise. Also you are using `posixpath` so I am assuming that you only want to treat `/`.

jakelishman replied:
Oh thanks for the catch - it's out-of-date documentation.  Originally I had written it to use `os.sep`, but then I realised that the whole directory handling in `autosummary` at the moment requires Unix-like paths and that it'd just make everything inconsistent for Windows builds to use different relative separators.

I've updated this documentation in 62b27c495.

jakelishman committed:
Correct toctree-directory documentation

AA-Turner commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -74,13 +74,23 @@ The :mod:`sphinx.ext.autosummary` extension does this in two parts:
+     relative to the directory of the source file, unless it begins with the
+     path separator ``/``, in which case it is relative to the documentation
+     root.  All paths should use the separator ``/`` regardless of the
+     operating system used for the build.
```
Currently inconsistent, but I'd prefer to say "source directory" rather than "root", as there are at least 4 direcories that users may configure. 
```suggestion
     as the file that contains the directive.
     The directory is interpreted as relative to the directory of the source file,
     unless it begins with ``/``, in which case it is interpreted as
     relative to the documentation source directory.
```

jakelishman replied:
Done in e4dd14fa7

AA-Turner commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -74,13 +74,23 @@ The :mod:`sphinx.ext.autosummary` extension does this in two parts:
 
+     .. versionadded:: 7.3
+
+        allowed setting a directory relative to the documentation root.
```
```suggestion
        Directories may now be set relative to the documentation source directory.
```

jakelishman replied:
Done in e4dd14fa7.

AA-Turner commented on sphinx/ext/autosummary/__init__.py:
```diff
@@ -234,9 +234,12 @@ def run(self) -> list[Node]:
-            dirname = posixpath.dirname(self.env.docname)
-
             tree_prefix = self.options['toctree'].strip()
+            if posixpath.isabs(tree_prefix):
```
Maybe consider using SEP rather than '/'. I'd take inpiration from how the ``:doc:`` role is implemented
```suggestion
            if tree_prefix.startswith('/'):
```

jakelishman replied:
Done in cdca8ed26.

I didn't use `os.sep` because it would cause a little bit of inconsistency with the rest of Autosummary on Windows, which uses `posixpath` for manipulations regardless of root platform, and `os.sep` on Windows is `\`.  There's `os.altsep` that's `/` on Windows, but I didn't immediately want to go that root, because I was a bit concerned that there'd be knock-on effects from the rest of the module still using `posixpath` and thought that changing the path handling more completely might want to be a different PR if there was a reason to do it.

AA-Turner replied:
Sorry; SEP is a constant that we define -- see ``osutil``.

A

jakelishman replied:
Ah sorry, I misunderstood.  That seems much preferable - the magic `"/"` constant didn't feel great. Changed in ~879b6b259~ 57fcb86e3 (because I forgot to run lint...).

AA-Turner replied:
(Aside: nice to see a fellow Leodensian!)

By way of an apology, autosummary has a lot of accumulated cruft and I tend to avoid looking at it out of habit. I'll try and look at this soon, but do feel free to ping. I'm currently working on the Python 3.12 documentation ahead of release next week, so as hopefully will get to this no later than mid-October.

Thanks,
Adam

jakelishman replied:
No worries at all, thanks - I'm in no huge rush.  If the only thing holding this up now is just your workload, I can just put it aside for a month and give it a bump if needed then.

I'm actually separately pushing to reduce our codebase's dependence on `autosummary` for similar reasons, and because (imo) it can be easy to fall into not-so-great documentation-structure habits if it's overused, which we certainly do.

(Yeah, it's not so common to see northerners around!  I grew up in York, went down south for undergrad and PhD - looks like you're similar - and now that's all done so I'm back.)

AA-Turner commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -629,7 +640,9 @@ def find_autosummary_in_lines(
                 toctree = m.group(1)
-                if filename:
+                if base_path is not None and posixpath.isabs(toctree):
+                    toctree = os.path.join(base_path, posixpath.relpath(toctree, "/"))
```
Similarly I'd prefer to be explicit that we're looking for a prefix of ``'/'``; semantically this is slightly different from normal path manipulations

jakelishman replied:
Done in cdca8ed26.

jakelishman committed:
Check for `/`-prefixed paths explicitly

Autosummary deals exclusively with URL-like path manipulations (i.e.
with an explicit `/`) rather than host-OS-specific ones.

jakelishman committed:
Fix up documentation wording

Co-authored-by: Adam Turner <9087854+aa-turner@users.noreply.github.com>

jakelishman committed:
Use `SEP` in place of magic literal for root marker
